metricbeat/module/zookeeper/mntr/mntr_test.go

```diff
@@ -0,0 +1,616 @@
+zk_p95_sync_processor_queue_flush_time_ms	3
+zk_p99_sync_processor_queue_flush_time_ms	3
+zk_p999_sync_processor_queue_flush_time_ms	3
+zk_avg_write_commitproc_time_ms	3.3333
+zk_min_write_commitproc_time_ms	1
+zk_max_write_commitproc_time_ms	6
+zk_cnt_write_commitproc_time_ms	6
+zk_sum_write_commitproc_time_ms	20
+zk_p50_write_commitproc_time_ms	4
+zk_p95_write_commitproc_time_ms	6
+zk_p99_write_commitproc_time_ms	6
+zk_p999_write_commitproc_time_ms	6
+zk_avg_10_learner_handler_qp_size	0.1429
+zk_min_10_learner_handler_qp_size	0
+zk_max_10_learner_handler_qp_size	2
+zk_cnt_10_learner_handler_qp_size	21
+zk_sum_10_learner_handler_qp_size	3
+zk_avg_zookeeper_read_per_namespace	212.0
+zk_min_zookeeper_read_per_namespace	212
+zk_max_zookeeper_read_per_namespace	212
+zk_cnt_zookeeper_read_per_namespace	1
+zk_sum_zookeeper_read_per_namespace	212
+zk_avg_v1_read_per_namespace	430.9095
+zk_min_v1_read_per_namespace	71
+zk_max_v1_read_per_namespace	9951
+zk_cnt_v1_read_per_namespace	199
+zk_sum_v1_read_per_namespace	85751
+zk_avg_zookeeper_write_per_namespace	141.3333
+zk_min_zookeeper_write_per_namespace	136
+zk_max_zookeeper_write_per_namespace	144
+zk_cnt_zookeeper_write_per_namespace	3
+zk_sum_zookeeper_write_per_namespace	424
+zk_avg_v1_write_per_namespace	2464.5476
+zk_min_v1_write_per_namespace	50
+zk_max_v1_write_per_namespace	11669
+zk_cnt_v1_write_per_namespace	126
+zk_sum_v1_write_per_namespace	310533
+zk_avg_10_learner_handler_qp_time_ms	1.0
+zk_min_10_learner_handler_qp_time_ms	1
+zk_max_10_learner_handler_qp_time_ms	1
+zk_cnt_10_learner_handler_qp_time_ms	1
+zk_sum_10_learner_handler_qp_time_ms	1
+zk_p50_10_learner_handler_qp_time_ms	1
+zk_p95_10_learner_handler_qp_time_ms	1
+zk_p99_10_learner_handler_qp_time_ms	1
+zk_p999_10_learner_handler_qp_time_ms	1`
+
+func TestEventMapping(t *testing.T) {
+
+	type TestCase struct {
+		Version    string
+		MntrSample string
+	}
+
+	mntrSamples := []TestCase{{"3.5.3", mntrTestInputZooKeeper35}, {"3.7.0", mntrTestInputZooKeeper37}}
+
+	logger := logp.NewLogger("mntr_test")
+
+	for i, sample := range mntrSamples {
+		versionMsg := fmt.Sprintf("With ZooKeeper [%s]", sample.Version)
+
+		reporter := &mbtest.CapturingReporterV2{}
+
+		eventMapping(fmt.Sprint(i), bytes.NewReader([]byte(sample.MntrSample)), reporter, logger)
+
+		assert.Empty(t, reporter.GetErrors(), versionMsg)
+
+		events := reporter.GetEvents()
+		assert.Len(t, events, 1, versionMsg)
+
+		event := events[len(events)-1]
+
+		// Check leader fields
+		assert.Equal(t, event.MetricSetFields["learners"], event.MetricSetFields["followers"], versionMsg)
+		assert.Equal(t, int64(1), event.MetricSetFields["learners"], versionMsg)
+
+		// Check latency fields
+		latencyFields := event.MetricSetFields["latency"].(common.MapStr)
+		assert.Equal(t, float64(29), latencyFields["max"], versionMsg)
+		if sample.Version != "3.5.3" {
+			// Float precission is only provided from ZK 3.6
+			assert.Equal(t, float64(0.5714), latencyFields["avg"], versionMsg)
```

Comment:
You could add a map of fields to compare in each `TestCase`, so you don't need conditional tests here.

It would be something like this in the definition:
```
	type TestCase struct {
		Version    string
		MntrSample string
		ExpectedValues map[string]interface{}
	}

	mntrSamples := []TestCase{
		{"3.5.3", mntrTestInputZooKeeper35, map[string]interface{
			"learners": int64(1),
			"latency.min": float64(0),
		}},
		{"3.7.0", mntrTestInputZooKeeper37, map[string]interface{
			"learners": int64(1),
			"latency.min": float64(0),
			"latency.avg": float64(0.5714),
		}},},
	}
```
And a loop here checking the fields.
```
for k, v := range sample.ExpectedValues {
        found, _ := event.MetricSetFields.Get(k)
	assert.Equal(t, v, found)
}
```

Reply:
Thanks @jsoriano , I thought about this but I wanted to keep it simple. For example: Comparing flattened maps for `latency` to nested maps would require some extra effort.

I am giving it a try anyway.

Reply:
Ahh nice! I had however already implemented a nested reports assertion function in 77212ad84c4dd198fe91de4c2f16aa94a5c90335

I'd like to keep it because I think it is easier to read nested maps than flattened ones (DRY :sweat_smile: ). With it, they look like:

```go
{
			"3.7.0",
			mntrTestInputZooKeeper37,
			common.MapStr{
				"learners":  int64(1),
				"followers": int64(1),
				"latency": common.MapStr{
					"max": float64(8),
					"avg": float64(0.5714),
					"min": float64(0),
				},
			},
		},
```

And the nested assertions checker looks like:

```go
func assertExpectations(t *testing.T, expectations common.MapStr, report common.MapStr, message ...string) {
	for key, expectation := range expectations {
		assert.Contains(t, report, key, message)
		switch expectation := expectation.(type) {
		case common.MapStr:
			nestedReport, _ := report.GetValue(key)
			assert.IsType(t, nestedReport, report, message)
			assertExpectations(t, expectation, nestedReport.(common.MapStr), message...)
		default:
			reportValue, _ := report.GetValue(key)
			assert.Equal(t, expectation, reportValue, message)
		}
	}
}
```